Assay import support for results domain file fields

cnathe opened the pull request:
#### Rationale
This story makes it easier to import an assay run to the app in the case where the assay results domain has File field types. Before this, you had to individually and manually add files to results after importing/creating the run. Now I can make the associations in bulk and upload the results files while importing a run through the app UI.

#### Related Pull Requests
- https://github.com/LabKey/labkey-api-js/pull/176
- https://github.com/LabKey/labkey-ui-components/pull/1501
- https://github.com/LabKey/labkey-ui-premium/pull/420
- https://github.com/LabKey/platform/pull/5533
- https://github.com/LabKey/limsModules/pull/302
- https://github.com/LabKey/commonAssays/pull/763

#### Changes
- FileAttachmentForm to send back updatedFiles map onFileRemoval
- FileAttachmentContainer to support includeDirectoryFiles to recursively include files in a dropped directory
- FileAttachmentContainer to support error check for total file size
- RunDataPanel update to show FileAttachmentForm for results domain file fields
- AssayWizardModel.prepareFormData to include resultsFiles for import run API post and update progress bar estimate and message
- uploadAssayRunFiles fix for maxFileSize to not include batchFiles or runFiles as they are unrelated
- only shown in assay import UI for LKB at this time

cnathe committed:
FileAttachmentForm to send back updatedFiles map onFileRemoval

cnathe committed:
RunDataPanel update to show FileAttachmentForm for results domain file fields

cnathe committed:
AssayWizardModel.prepareFormData to include resultsFiles for import run API post

cnathe committed:
3.45.0-fb-assayResultsFiles.2

cnathe committed:
uploadAssayRunFiles fix for maxFileSize to not include batchFiles or runFiles as they are unrelated

cnathe committed:
3.45.0-fb-assayResultsFiles.3

cnathe committed:
Merge branch 'develop' into fb_assayResultsFiles

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
Merge branch 'develop' into fb_assayResultsFiles

# Conflicts:
#	packages/components/releaseNotes/components.md

cnathe committed:
3.46.2-fb-assayResultsFiles.0

cnathe committed:
FileAttachmentContainer to support includeDirectoryFiles to recursively include files in a dropped directory

cnathe committed:
RunDataPanel update to file drop target labels

cnathe committed:
snapshot update for new prop

cnathe committed:
AssayImportPanels update progress bar estimate and message

cnathe committed:
FileAttachmentContainer to support includeDirectoryFiles to recursively include files in a dropped directory

cnathe committed:
FileAttachmentContainer to support error check for total file size

cnathe committed:
RunDataPanel help tips for file drop area labels

cnathe committed:
jest snapshot updates

cnathe committed:
3.46.2-fb-assayResultsFiles.1

cnathe committed:
null check

cnathe committed:
3.46.2-fb-assayResultsFiles.2

cnathe committed:
small css tweak

cnathe committed:
Merge branch 'develop' into fb_assayResultsFiles

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.46.3-fb-assayResultsFiles.0

cnathe committed:
jest snapshot update

cnathe committed:
only shown in assay import UI for LKB at this time (add isLIMSEnabled() helper)

cnathe committed:
3.46.3-fb-assayResultsFiles.1

cnathe committed:
npm run lint-branch-fix

cnathe committed:
Merge branch 'develop' into fb_assayResultsFiles

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.46.4-fb-assayResultsFiles.0

cnathe committed:
Merge branch 'develop' into fb_assayResultsFiles

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.46.5-fb-assayResultsFiles.0

labkey-nicka commented on packages/components/src/internal/AssayDefinitionModel.ts:
```diff
@@ -297,4 +297,11 @@ export class AssayDefinitionModel extends ImmutableRecord({
     }
+
+    domainFileColumnNames(type: AssayDomainTypes): string[] {
+        return this.getDomainByType(type)
```
`this.getDomainByType(type)` can return `undefined`. Add defensive checks here and return an empty array or undefined.

labkey-nicka commented on packages/components/src/internal/AssayDefinitionModel.ts:
```diff
@@ -297,4 +297,11 @@ export class AssayDefinitionModel extends ImmutableRecord({
         return columns.asImmutable();
     }
+
+    domainFileColumnNames(type: AssayDomainTypes): string[] {
```
nit: Perhaps this should be `getDomainFileColumns` and then let the caller map out the names? I guess I'm thinking this should be similar to `getDomainColumns(type: AssayDomainTypes)`. Something like this:

```tsx
getDomainFileColumns(type: AssayDomainTypes): QueryColumn[] {
    return this.getDomainColumns(type)
        .filter(col => col.isFileInput)
        .toArray();
}
```

Then it wouldn't need to be defensive (`getDomainColumns` already is) and we leave it up to the caller if they need more column information.

labkey-nicka commented on packages/components/src/internal/components/assay/RunDataPanel.tsx:
```diff
@@ -38,12 +38,34 @@ import { Alert } from '../base/Alert';
 
 import { getContainerFilterForLookups } from '../../query/api';
 
+import { AssayDomainTypes } from '../../AssayDefinitionModel';
```
nit: There is dead code in this file. `onTextChange` and `clearText` can be removed.

labkey-nicka commented on packages/components/src/internal/components/assay/RunDataPanel.tsx:
```diff
@@ -38,12 +38,34 @@ import { Alert } from '../base/Alert';
+const RESULTS_FILE_SIZE_LIMITS = Map<string, FileSizeLimitProps>({
+    all: {
+        totalSize: {
+            value: 209715200,
```
This value is worthy of a comment

cnathe replied:
Done

labkey-nicka commented on packages/components/src/internal/components/assay/RunDataPanel.tsx:
```diff
@@ -231,11 +255,29 @@ export class RunDataPanel extends PureComponent<Props, State> {
+        const fileColumnNames = isLIMSEnabled()
+            ? wizardModel.assayDef.domainFileColumnNames(AssayDomainTypes.RESULT)
+            : undefined;
+        const assayFileFieldTooltip = (
```
nit: Would be great if `<LabelOverlay/>` accepted `children` as the `content` so this could just be done inline. A bit of wasted processing going on here.

cnathe replied:
yeah, not sure why `content` was added as a prop instead of just using `children`. I made that change, looks like there wasn't many usages.

labkey-nicka replied:
Cool. Thanks for doing that. 

nit: For this usage in particular the `assayFileFieldTooltip` could now just be inlined and would not be processed when there are no file columns.

labkey-nicka commented on packages/components/src/internal/components/assay/AssayWizardModel.ts:
```diff
@@ -96,6 +99,24 @@ export class AssayWizardModel
         return this.attachedFiles.valueSeq().toList();
     }
 
+    getTotalAttachedFilesSize(): number {
```
These methods can avoid proliferating `immutable` objects by just returning arrays. All the use cases either convert them to arrays or check for size neither of which require immutable structures.

```tsx
getAttachedFiles(): File[] {
    return this.attachedFiles.valueSeq().toArray();
}

getTotalAttachedFilesSize(): number {
    return this.attachedFiles.reduce((totalSize, file) => totalSize + file.size, 0);
}

getResultsFiles(): File[] {
    return this.resultsFiles.valueSeq().toArray();
}

getTotalResultsFilesSize(): number {
    return this.resultsFiles.reduce((totalSize, file) => totalSize + file.size, 0);
}
```

labkey-nicka commented on packages/components/src/internal/components/assay/AssayWizardModel.ts:
```diff
@@ -16,6 +16,7 @@ export interface AssayUploadOptions extends AssayDOM.ImportRunOptions {
     dataRows?: any; // Array<any>
     maxFileSize?: number;
     maxRowCount?: number;
+    resultsFiles?: any[];
```
Follow-on from the comment in `@labkey/api` PR that this could/should be `File[]`.

labkey-nicka commented on packages/components/src/internal/components/assay/AssayImportPanels.tsx:
```diff
@@ -664,8 +681,8 @@ class AssayImportPanelsBody extends Component<Props, State> {
-        if (data.files && data.files.length > 0) {
-            return data.files[0].size * 0.2;
+        if (data.files?.length > 0) {
+            return model.getTotalAttachedFilesSize() / 5 + model.getTotalResultsFilesSize() / 100_000;
```
This line deserves a comment given the constants.

labkey-nicka commented on packages/components/src/internal/components/assay/RunDataPanel.tsx:
```diff
@@ -56,9 +78,11 @@ interface Props {
     maxRows?: number;
-    onFileChange: (attachments: Map<string, File>) => any;
-    onFileRemoval: (attachmentName: string) => any;
+    onDataFileChange: (attachments: Map<string, File>) => any;
```
nit: Switch all these `=> any` to `=> void`.

labkey-nicka commented on packages/components/src/internal/components/files/FileAttachmentContainer.tsx:
```diff
@@ -92,10 +98,59 @@ export class FileAttachmentContainer extends React.Component<
         });
     };
 
+    getFilesFromDirectory = (files, entry, scope, callback): void => {
```
This method is lacking in typings. Was this copied from somewhere else? Some left-hand typings do help a bit but also it is helpful here to use the "is" pattern to recognize types at run time. Here are a couple of utilities:

```tsx
const isDirectoryEntry = (entry: FileSystemEntry): entry is FileSystemDirectoryEntry => entry.isDirectory;
const isFileEntry = (entry: FileSystemEntry): entry is FileSystemFileEntry => entry.isFile;
```

I was able to get it pretty much fully typed:

```tsx
getFilesFromDirectory = (
    files: Record<string, File>,
    entry: FileSystemDirectoryEntry,
    scope: FileAttachmentContainer,
    callback: () => void
): void => {
    const dirReader: FileSystemDirectoryReader = entry.createReader();

    const entriesReader: FileSystemEntriesCallback = (scope => {
        return entries => {
            for (let i = 0; i < entries.length; i++) {
                const _entry = entries[i];
                if (isFileEntry(_entry)) {
                    scope.fileCbCount++;
                    _entry.file(file => {
                        scope.fileCbCount--;

                        // ignore hidden files
                        if (file.name.substring(0, 1) !== '.') {
                            if (files[file.name]) {
                                scope.setState({
                                    warningMsg:
                                        'Duplicate files were uploaded. Only the last file provided for the duplicate name will be included in the file set.',
                                });
                            }
                            // @ts-expect-error fullPath is not a property of File in lib.dom.ts
                            file.fullPath = entry.fullPath + '/' + file.name;
                            files[file.name] = file;
                        }

                        if (scope.dirCbCount === 0 && scope.fileCbCount === 0) {
                            callback();
                        }
                    });
                } else if (isDirectoryEntry(_entry)) {
                    scope.getFilesFromDirectory(files, _entry, scope, callback);
                }
            }

            // readEntries only reads 100 files in a batch for Chrome
            if (entries.length >= 100) {
                scope.dirCbCount++;
                dirReader.readEntries(entriesReader, console.error);
            }

            scope.dirCbCount--;
            if (scope.dirCbCount === 0 && scope.fileCbCount === 0) {
                callback();
            }
        };
    })(scope);

    scope.dirCbCount++;
    dirReader.readEntries(entriesReader, console.error);
};
```

cnathe replied:
Awesome. Thanks. This was copied (and extended a bit) from the LABKEY.internal.ZipLoad.getFilesFromDirectory function in ZipLoad.js.

labkey-nicka commented on packages/components/src/public/files/FileAttachmentForm.tsx:
```diff
@@ -41,20 +41,22 @@ interface FileAttachmentFormProps {
+    sizeLimitsHelpText?: ReactNode;
     showAcceptedFormats?: boolean;
     allowDirectories?: boolean;
+    includeDirectoryFiles?: boolean;
```
Careful when linting here as these inline comments can get thrown out of order. I typically hand edit these out of order property listings.

cnathe committed:
Merge branch 'develop' into fb_assayResultsFiles

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.47.0-fb-assayResultsFiles.0

cnathe committed:
CR feedback (patch from Nick):
- linting
- domainFileColumnNames -> getDomainFileColumns
- add comment re: progress bar estimate
- AssayWizardModel get files helpers to not return immutable lists
- LabelOverlay to use `children` instead of `constent` prop
- FileAttachmentContainer.getFilesFromDirectory typings

cnathe committed:
3.47.0-fb-assayResultsFiles.1

cnathe committed:
3.47.0-fb-assayResultsFiles.2

cnathe committed:
jest tests part 1

cnathe committed:
RunDataPanel test convert to RTL and add test case for File Fields Data drop target

cnathe committed:
FileAttachmentContainer test convert to RTL and update for file listing display changes

cnathe committed:
FileAttachmentContainer test convert to RTL and update for file listing display changes

cnathe committed:
Fix for import button initially disabled on assay run re-import

cnathe committed:
3.47.0-fb-assayResultsFiles.3

cnathe committed:
Merge branch 'develop' into fb_assayResultsFiles

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.48.0-fb-assayResultsFiles.0

cnathe committed:
Merge branch 'develop' into fb_assayResultsFiles

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.48.1-fb-assayResultsFiles.0

cnathe committed:
Merge branch 'develop' into fb_assayResultsFiles

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.49.0-fb-assayResultsFiles.0

cnathe committed:
Merge branch 'develop' into fb_assayResultsFiles

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.50.0-fb-assayResultsFiles.0

cnathe committed:
npm run lint-branch-fix

cnathe committed:
Update release notes with version number and release date

cnathe committed:
3.51.0

cnathe committed:
jest snapshot update
